flintrock/core.py

```diff
@@ -506,42 +506,66 @@ def get_java_major_version(client: paramiko.client.SSHClient):
+            # First line of the output is like: 'openjdk version "1.8.0_252"' or 'openjdk version "11.0.7" 2020-04-14'
             # Get the version string and strip out the first two parts of the
             # version as a tuple: (1, 8)
             if len(tokens) >= 3:
                 version_parts = tokens[2].strip('"').split(".")
                 if len(version_parts) >= 2:
-                    return tuple(int(part) for part in version_parts[:2])
+                    if version_parts[0] == "1":
+                        # Java 6, 7 or 8
+                        return version_parts[1]
+                    else:
+                        # Java 9+
+                        return version_parts[0]
         except SSHError:
             pass
 
     return None
 
 
-def ensure_java8(client: paramiko.client.SSHClient):
+def ensure_java(client: paramiko.client.SSHClient, jdk: str):
     host = client.get_transport().getpeername()[0]
     java_major_version = get_java_major_version(client)
 
-    if not java_major_version or java_major_version < (1, 8):
-        logger.info("[{h}] Installing Java 1.8...".format(h=host))
+    # sudo yum install -y java-1.8.0-openjdk
+    # sudo amazon-linux-extras install -y java-openjdk11
+    if not java_major_version or java_major_version < jdk:
+        logger.info("[{h}] Installing Java {j}...".format(h=host, j=jdk))
 
+        install_adoptopenjdk_repo(client)
+        java_package = "adoptopenjdk-{j}-hotspot".format(j=jdk)
         ssh_check_output(
             client=client,
             command="""
                 set -e
 
-                # Install Java 1.8 first to protect packages that depend on Java from being removed.
-                sudo yum install -y java-1.8.0-openjdk
+                # Set up the AdoptOpenJDK yum repo
+                sudo mv /tmp/adoptopenjdk.repo /etc/yum.repos.d/
+                
+                # Install Java first to protect packages that depend on Java from being removed.
+                sudo yum install -q -y {jp}
 
                 # Remove any older versions of Java to force the default Java to 1.8.
                 # We don't use /etc/alternatives because it does not seem to update links in /usr/lib/jvm correctly,
                 # and we don't just rely on JAVA_HOME because some programs use java directly in the PATH.
                 sudo yum remove -y java-1.6.0-openjdk java-1.7.0-openjdk
 
-                sudo sh -c "echo export JAVA_HOME=/usr/lib/jvm/jre >> /etc/environment"
+                # sudo alternatives --set java {jp}.x86_64
+                sudo sh -c "echo export JAVA_HOME=/usr/lib/jvm/{jp} >> /etc/environment"
                 source /etc/environment
-            """)
+            """.format(jp=java_package))
+
+
+def install_adoptopenjdk_repo(client):
+    """
+    Copies the adoptopenjdk.repo file to the /tmp directory on the client.
+    It must be sudo mv'd to /etc/yum.repos.d/ later.
```

Comment:
Why not do the `mv` as part of this function? Seems like it would be cleaner to have this setup happen in one go.

If you think these steps should be separate, then at the very least I would suggest returning the remote path from here and using that in `ensure_java()`, so that `ensure_java()` doesn't just assume where to find the script.